File: Leetcode/374.cpp
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        int low = 0;
        int high = n;
        while (true) {
            if (low > high) {
                return -1;
            }
            int mid = low + (high - low) / 2;
            int key = guess(mid);
            if (key == -1) {
                high = mid - 1;
            } else if (key == 1) {
                low = mid + 1;
            } else {
                return mid;
            }
        }
    }
};

File: Leetcode/374/374.go
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * func guess(num int) int;
 */

func guessNumber(n int) int {
	l := 1
	r := n
	var mid, temp int
	for l <= r {
		mid = l + ((r-l) >> 1)
		temp = guess(mid)
		if temp == 0 {
			return mid
		} else if temp > 0 {
			l = mid + 1
		} else {
			r = mid - 1
		}
	}
	return -1
}